src/nacl/secret.py

```diff
@@ -132,3 +132,121 @@ def decrypt(self, ciphertext, nonce=None, encoder=encoding.RawEncoder):
         )
 
         return plaintext
+
+
+class Aead(encoding.Encodable, StringFixer, object):
+    """
+    The AEAD class encrypts and decrypts messages using the given secret key.
+
+    Unlike :class:`~nacl.secret.SecretBox`, AEAD supports authenticating
+    non-confidential data received alongside the message, such as a length
+    or type tag.
+
+    Like :class:`~nacl.secret.Secretbox`, this class provides authenticated
+    encryption. An inauthentic message will cause the decrypt function to raise
+    an exception.
+
+    Likewise, the authenticator should not be mistaken for a (public-key)
+    signature: recipients (with the ability to decrypt messages) are capable of
+    creating arbitrary valid message; in particular, this means AEAD messages
+    are repudiable. For non-repudiable messages, sign them after encryption.
+
+    :param key: The secret key used to encrypt and decrypt messages
+    :param encoder: The encoder class used to decode the given key
+
+    :cvar KEY_SIZE: The size that the key is required to be.
+    :cvar NONCE_SIZE: The size that the nonce is required to be.
+    :cvar MACBYTES: The size of the authentication MAC tag in bytes.
+    :cvar MESSAGEBYTES_MAX: The maximum size of a message which can be
+                            safely encrypted with a single key/nonce
+                            pair.
+    """
+
+    KEY_SIZE = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_KEYBYTES
+    NONCE_SIZE = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES
+    MACBYTES = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_ABYTES
+    MESSAGEBYTES_MAX = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_MESSAGEBYTES_MAX
+
+    def __init__(self, key, encoder=encoding.RawEncoder):
+        key = encoder.decode(key)
+        if not isinstance(key, bytes):
+            raise exc.TypeError("AEAD must be created from 32 bytes")
+
+        if len(key) != self.KEY_SIZE:
+            raise exc.ValueError(
+                "The key must be exactly %s bytes long" % self.KEY_SIZE,
+            )
+
+        self._key = key
+
+    def __bytes__(self):
+        return self._key
+
+    def encrypt(self, plaintext, aad=b'', nonce=None, encoder=encoding.RawEncoder):
+        """
+        Encrypts the plaintext message using the given `nonce` (or generates
+        one randomly if omitted) and returns the ciphertext encoded with the
+        encoder.
+
+        .. warning:: It is **VITALLY** important that the nonce is a nonce,
```

Comment:
I kept that from `SecretBox`, but TBH I don't particularly like that wording?
Encouraging users to manage nonces on their own, seems to me much more dangerous than risking randomly-generated nonces colliding (especially since XChacha20 uses 192b nonces, so users are safe with up to 2⁹⁶ messages encrypted under the same key and random nonces)